src/storage/sqlite.js

```diff
@@ -97,18 +75,37 @@ export class SQLDatabase {
 // An absurd little workaround for expo-sqlite, or really
 // the @expo/websql library under it, being too eager to check a
 // transaction's queue and declare it complete.
-async function keepQueueLiveWhile(
-  tx: WebSQLTransaction,
-  f: () => void | Promise<void>,
-): Promise<void> {
+//
+// Also for it not handling errors in callbacks, so that an exception in a
+// transaction's application-level code causes it to commit (!) if the
+// transaction callback itself throws an exception, and to get the whole
+// database object stuck if a statement callback throws an exception.
```

Comment:
> Do you think there might be lingering cases to be handled where the whole DB object can get stuck? I wonder if we could get good logging to Sentry when this happens.

Yeah, good question. I'm not prepared to rule out that that could happen, but I think it would at least be a distinct issue -- I think this issue is completely resolved by what's in the wrapper now.

Specifically the issue is that in `node_modules/@expo/websql/lib/websql/WebSQLDatabase.js`, the `_running` property on the `WebSQLDatabase` instance gets set to true, blocking any new transactions, and won't get set back to false until `_onTransactionComplete` is called. Looking at its sibling `WebSQLTransaction.js`, it won't call that until the transaction's own `_running` property gets set back to false. And at the call sites of `batchTask.sqlCallback` and `batchTask.sqlErrorCallback`, if those throw an exception then that propagates right past where we were going to call the local `onDone`, which was the place that was supposed to set `._running` back to false.

In our wrapper, the callbacks passed to `executeSql` are always trivial and shouldn't throw; they just resolve or reject a promise. So their caller in `WebSQLTransaction.js` should always get through them and make it to `onDone`, so things don't get stuck. (If app code hits an error, that happens in the promise's callback (or further down the line), after we resolve or reject it and after we've returned to the WebSQLTransaction code to let it complete its work.) That leaves the danger it'll go ahead and commit when it shouldn't.

And that in turn is resolved by the `hold` loop keeping something in the queue, and putting something error-causing there when the transaction should fail.


> Does this happen when an error is thrown from within an error callback, not a success callback, and should we handle that as well?

Yeah, as described above, it does and we do. "A statement callback" was meant to cover both kinds of callbacks from `executeSql`, but could perhaps be more explicit.

Reply:
Added a link in that code comment back to this subthread.